### DISCUSSION TITLE
Writable tensor?

### DISCUSSION THREAD

**Discussion by qnzhou** [wjakob/nanobind#46] [Q&A] [answered]
I am wondering if it possible to expose C++ buffers as a writable tensor in python.  E.g. if I have the `Foo` class as
```C++
struct Foo {
    std::vector<int> tags = {1, 2, 3};
};
```

and the `tags` member variable is exposed in python using `nb::tensor<>` like the following.

```C++
using IntArray = nb::tensor<nb::numpy, int, nb::shape<nb::any>>;
nb::class_<Foo>(m, "Foo")
    .def_property("tags", 
    [](Foo& foo) {
        size_t shape[] = {foo.tags.size()};
        return IntArray(reinterpret_cast<void*>(foo.tags.data()), 1, shape);
    }, [](Foo& foo, IntArray tensor) {
        ...
    });
```

In python, the numpy array corresponding to `tags` appears to be not writable.
```python
a = Foo()
print(a.tags.flags)
#  C_CONTIGUOUS : True
#  F_CONTIGUOUS : True
#  OWNDATA : False
#  WRITEABLE : False
#  ALIGNED : True
#  WRITEBACKIFCOPY : False
```

Is there a way of creating writable tensor such that in place update like `a.tags[0] = 10` is allowed?

**Comment by brettc** [comment, marked as answer]
As @qnzhou said in #53, this looks like a limitation of using `from_dlpack()` and is by design. A related issue is the one mentioned in #42, where you cannot send a read-only numpy array to c++, though it looks like this might be worked around in nanobind.

So working through dlpack currently puts some constraints on passing tensors.
1. You can't pass writeable tensors from c++ -> python.
2. You can't pass read-only tensors from python -> c++.

I see that the reimplementation of the numpy interface is "open for discussion". @wjakob, is it worthwhile submitting a PR that ports the numpy code from pybind11? I could give this a go, though it might be a challenge given my limited understanding of the internals.

**Reply by wjakob** [reply]
I don't think it will be a good idea to port things from pybind11 in this case. FWIW, to support older NumPy versions (which didn't support the DLPack interface), I actually included an alternative path that goes through the classic buffer protocol. You should see this when looking at the code in `nb_tensor.cpp`. Could you see if forcing that fixes your use case?

**Reply by brettc** [reply]
Aha! You've implemented the buffer protocol for the dltensor type. I can make the returned buffer writeable by simply hacking this line:
https://github.com/wjakob/nanobind/blob/43a96c537824b4a8b5202e166c1bff239a7e547e/src/tensor.cpp#L587
to:
```python
            o = package.attr("asarray")(o);
```
Of course, this breaks everything other than numpy, but it does show that the code to do this is available. 

The question is how to best let the user choose this path. It seems to me that treating numpy as a special case where there is a choice about the whether the buffer is RW or RO might fix both cases I mention above. As I mentioned in #42, modifying the flags on the `PyOjbect_GetBuffer` can allow RO buffers to be sent from python.

 Is adding some `writeable` flag to the nb::tensor possible? I'd be happy to submit a PR, but would prefer getting some guidance first.

**Reply by qnzhou** [reply]
Taking one step back, [this numpy issue](https://github.com/numpy/numpy/issues/20337) confirms that @brettc's observation is by design from numpy's side.  @rgommers mentioned

> I think importing readonly is nice: it's quite safe, should not interfere with most normal user code, and in case it does the user can consciously flip the writeable flag.

However, for buffers exposed via `nb::tensor`, flipping the write flag causes the following error:
```python
>>> a.tags.setflags(write=True)
ValueError: cannot set WRITEABLE flag to True of this array
```
I am not sure why it is the case...

**Reply by wjakob** [reply]
I would be fine with having a `nb::writable` annotation to force the buffer protocol path in NumPy.

**Reply by qnzhou** [reply]
An interesting observation is that with older numpy (e.g. v1.21.5), the numpy array that wraps around a nanobind tensor is writeable out of the box.  With newer version of numpy (e.g. v1.24.1), it is not, and the writeable flag cannot be set.

**Reply by wjakob** [reply]
That's likely the cutoff where nanobind switches over from talking to numpy using the buffer protocol vs DLPack.

**Reply by qnzhou** [reply]
FYI, this issue seems to be fixed in v1.0.0 release with `ndarray` replacing `tensor`.  I still have a few questions:

* What exact change fixed this issue? (Just curious).
* Now, is there a way to bind a read-only buffer with `ndarray` with `writable` bit off?  Ideally without `const_cast`.

**Reply by wjakob** [reply]
For #1, see the changelog. For #2, unfortunately not yet. `nb::ndarray` does not include a read-only bit yet, help on this front is appreciated.
